src/recompiler.ts

```diff
@@ -0,0 +1,262 @@
+                changed.add(name);
+            }
+        }
+
+        return changed;
+    }
+
+    // Compile changed contracts
+    private async compileChangedContracts(
+        newCacheEntry: Record<string, ContractData>,
+        changed: Set<string>
+    ): Promise<void> {
+        for (const changedContract of changed) {
+            const entry = newCacheEntry[changedContract];
+            const compileArguments: TaskArguments = {
+                paths: [changedContract],
+                disableHintValidation: entry?.disableHintValidation,
+                accountContract: entry?.accountContract,
+                carioPath: entry?.cairoPath
+            };
+
+            await starknetCompileAction(compileArguments, this.hre);
+        }
+    }
+
+    // Updated set with changed and new contracts
+    private async updateSet(
+        cache: Record<string, ContractData>,
+        newCacheEntry: Record<string, ContractData>,
+        changed: Set<string>
+    ): Promise<Set<string>> {
+        for (const contractName in newCacheEntry) {
+            // Add new contracts that are not in cache before
+            if (!cache[contractName]) {
+                changed.add(contractName);
+                continue;
+            }
+
+            // Add contracts that contiain a change in content
+            if (newCacheEntry[contractName].contentHash !== cache[contractName].contentHash) {
+                changed.add(contractName);
+            }
+        }
+
+        // Remove deleted sources from old cache by overwriting it
+        this.setCache(newCacheEntry);
+        await this.saveCache();
+
+        return changed;
+    }
+
+    // Handles cache on Starknet cli calls
+    public async handleCache(): Promise<void> {
+        // If recompile is not enabled, do nothing
+        if (!this.hre.userConfig?.starknet?.recompile) return;
+
+        const paths = this.hre.config.paths;
+        try {
+            const oldCache = await this.getCache();
+            const newCacheEntry = await this.getContractHash(paths);
+            const changedContracts = await this.checkArtifacts(paths, newCacheEntry);
+            const updatedSet = await this.updateSet(oldCache, newCacheEntry, changedContracts);
+            await this.compileChangedContracts(newCacheEntry, updatedSet);
+        } catch (error) {
+            // If there is an error, do not recompile
+            console.error(error);
+            process.exit(1);
+        }
+    }
+
+    // Updates cache with new contract and artifacts
+    public async updateCache(
+        args: TaskArguments,
+        file: string,
+        output: string,
+        abi: string,
+        cairoPath?: string
+    ): Promise<void> {
+        const oldCache = await this.getCache();
+        const newCacheEntry = await this.getCacheEntry(file, output, abi, cairoPath, args);
+        this.cache = this.getUpdatedCache(oldCache, newCacheEntry);
+        this.saveCache();
```

Comment:
I think we have to many filesystem calls which can be very slow. Can we first update every entry and then save cache in `starknetCompileAction`?

Reply:
https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/106/commits/077b8b8d734c1db29acd47f09b6ca9e338804860